### PR TITLE
CSF: Use `__orderedExports` in loader if provided 

### LINKED ISSUE #9136: CSF story order depends on webpack and breaks with vanilla es module imports
**Describe the bug**
The order of stories within the code of an `index.stories.js` file (in CSF format) determines the order of the storybook menu.

Consider the following example
```js
// orderedNamedExports.js
export const b = 'BB';
export const a = 'AA';
export const c = 'CC';

// start.js
import('./orderedNamedExports.js').then(module => {
  console.log(Object.keys(module));
});
```

When loading this file by going through webpack it will result in
```js
// (3) ["b", "a", "c"]
```

However, when using a browser native es module import the modules will be sorted.
```js
// (3) ["a", "b", "c"]
```

It is specified in the es module spec https://tc39.es/ecma262/#sec-modulenamespacecreate

> 7. Let sortedExports be a new List containing the same values as the list exports where the values are ordered as if an Array of the same values had been sorted using Array.prototype.sort using undefined as comparefn.

There are also part of the official platform tests https://github.com/tc39/test262/blob/master/test/language/module-code/namespace/internals/own-property-keys-sort.js

it sort of all started with this tweet so if you are interested there are more details
https://twitter.com/daKmoR/status/1204877164233015297?s=20

As this order determines the menu order it will result in "wrong menus" if you are not using webpack[1] - (and possible also when doing production builds with the upcoming es module flag in webpack v5)

**To Reproduce**
1. have a story file like
```js
import { html } from '../../index.js';

export default {
  title: 'Has order Paragraph-Heading',
};

export const paragraph = () =>
  html`
    <p>First Story</p>
  `;

export const heading = () =>
  html`
    <h1>Second Story</h1>
  `;
```
2. using webpack results in the following menu
  - Paragraph
  - Heading
3. using vanilla es module imports results in the following menu
  - Heading
  - Paragraph

**Expected behavior**
Have storybook have the same story order when using webpack and when using es modules.

**strawman proposal**
We could export a special `__exportOrder` array from within story files. This could be automatically done by a loader.

possible addition for the example above
```js
// 1.
export __exportOrder = ['paragraph', 'heading'];

// 2.
export __orderedExports = [paragraph, heading];
```

with 1. you would sort the * imports via this array and with 2. you would ignore all other imports and just use this ordered array contain all exports.

Personally I think 2. sounds better

What do you all think?

---
[1]: > A quick check shows the problem is not limited to webpack; Babel & TypeScript also make no attempt to sort exports. [tweet](https://twitter.com/robpalmer2/status/1205066610769354752?s=20)

### PR DESCRIPTION
**Pull request by daKmoR**
closes: https://github.com/storybookjs/storybook/issues/9136

## What I did

prefer a user/loader provided `__orderedExports` object if supplied as es module exports are ordered alphabetically

This will have 0 effects if that export is not used/provided.

And you should only provide it you are using native es modules (which have alphabetically ordered exports vs webpack source code ordered exports)

## How to test

- use storybook without webpack and provide this export

### CODE REVIEW

**Comment by daKmoR** [comment]
oh no a used an uppercase character in the branch 🙈 
therefore the react native smoke test is failing - shall I close and reopen with a different branch?

**Review by shilman — APPROVED** [review]
Looks great! @ndelangen @tmeasday @Hypnosphi can you take a look?

File: lib/core/src/client/preview/start.js [hunk 1]
```diff
@@ -379,7 +379,10 @@ export default function start(render, { decorateStory } = {}) {
         );
       }
 
-      const { default: meta, ...exports } = fileExports;
+      const { default: meta, __orderedExports, ...namedExports } = fileExports;
+      // prefer a user/loader provided `__orderedExports` object if supplied as es module exports
+      // are ordered alphabetically - see https://github.com/storybookjs/storybook/issues/9136
+      const exports = __orderedExports || namedExports;
```
Comment (Hypnosphi) [inline lib/core/src/client/preview/start.js]: Does js object literal guarantee `Object.keys()` order nowadays? Or would it be safer to use an array instead?

Actually, I think we can use the existing [`іncludeStories`](https://storybook.js.org/docs/formats/component-story-format/#non-story-exports) API for that, instead of introducing a new one

```js
import React from 'react';
import MyComponent from './MyComponent';
import someData from './data.json';

export default {
  title: 'MyComponent',
  component: MyComponent,
  // should also affect the order
  includeStories: ['SimpleStory', 'ComplexStory']
}

export const simpleData = { foo: 1, bar: 'baz' };
export const complexData = { foo: 1, { bar: 'baz', baz: someData }};

export const SimpleStory = () => <MyComponent data={simpleData} />;
export const ComplexStory = () => <MyComponent data={complexData} />;
```

Comment (shilman) [reply lib/core/src/client/preview/start.js]: Good point @Hypnosphi, tho much harder to implement IMHO

Comment (daKmoR) [reply lib/core/src/client/preview/start.js]: as far as I can tell `includeStories` and `__orderedExports` serve a different purpose?

- `includeStories` is handcrafted by users to define specific stories that should be included (and in which order)
- `__orderedExports` should be auto-generated and should include always all stories (can still be filtered with exclude/includeStories

> Does js object literal guarantee Object.keys() order nowadays?

not in the spec but in all practical cases yes. (also the current * import is an Object as well 😬)

We can have an array but it will be more complicated - see https://github.com/storybookjs/storybook/issues/9136#issuecomment-566782481 for an example

imho we should not "misuse"  `includeStories` for it... also there are more questions like what if the users provide includeStories and I want to use it to export the default order? what wins? it would move certain logic into the CSF story files (or its generation/postprocessing) imho not the right place for that logic

Comment (shilman) [reply lib/core/src/client/preview/start.js]: i agree. it's too complicated to overload `includeStories`!

**Comment by tmeasday** [comment]
So long term we would expect that users would have to use a babel plugin or equivalent to get this `__orderedExports` object in CSF? (It sounds like webpack itself may move to this new behaviour: https://twitter.com/wSokra/status/1205183567422201857).

What a pain. I guess this solution is the best we can do?

Here's another approach I thought of, that's kind of crappy but would work:

We discussed having a simple boxing function to create stories once upon a time:
```js
import { story } from '@storybook/csf';

export storyOne = story(() => <Component />, { metadata });
```

We could make the `story()` function keep an internal counter of "storyNumber" and add that as a part of the storyMetadata. A big change though.

Just wanted to write that down.

**Comment by daKmoR** [comment]
TLDR;
- will only be needed if source order matters (like in storybook) but not for example when using in tests
- is not needed in storybook with webpack right now (and for the foreseeable future)

Details:
you will only need this if the order actually matters e.g. if you use it for tests or for displaying them inside of other systems it is not needed.

Only if you want to display a menu based on the source order of the code then you need to do something. And in any case, it really depends on your setup. For storybook with webpack right now nothing is needed (at some far distance future maybe a webpack plugin or a loader).

If you are using vanilla es modules then you need to auto-generate this __orderedExports. Every system will need to find its own way - We will be using [es-module-lexer](https://github.com/guybedford/es-module-lexer) (a lexer written in c and made available to the browser via wasm 🤗) to generate it on the fly via a transformer (on stories.js and stories.mdx files) with hopefully very little performance impact

Regarding
```js
import { story } from '@storybook/csf';

export storyOne = story(() => <Component />, { metadata });
```
this could indeed solve it as well but it puts the complexity on the users side...
imho retaining the source order should be an implementation detail - especially as it can be solved via a transformer or loader

File: lib/core/src/client/preview/start.js (hunk 1 above)
Comment (Hypnosphi) [reply lib/core/src/client/preview/start.js]: Ok, can it at least share the same API?

```js
export default {
  title: 'MyComponent',
  component: MyComponent,
  storiesOrder: ['SimpleStory', 'ComplexStory']
}
```

Comment (Hypnosphi) [reply lib/core/src/client/preview/start.js]: > much harder to implement

Not really, you just need to iterate over `meta.storiesOrder` instead of `Object.keys(exports)`

**Comment by tmeasday** [comment]
> this could indeed solve it as well but it puts the complexity on the users side...
imho retaining the source order should be an implementation detail - especially as it can be solved via a transformer or loader

I agree with this sentiment but I just dream of a day where we can drop babel and webpack and all source transformation tools and just load things natively. This seems like a pretty big roadbump in getting there :/

File: lib/core/src/client/preview/start.js (hunk 1 above)
Comment (daKmoR) [reply lib/core/src/client/preview/start.js]: the complexity is not on the iteration part it is on the loader/processor part

e.g. take this input
```js
export default {
  title: 'MyComponent',
  component: MyComponent
}

export const paragraph = () =>
  html`
    <p>First Story</p>
  `;

export const heading = () =>
  html`
    <h1>Second Story</h1>
  `;
```

now if you want to add the order all you need to do is add on extra line at the end of the file
```js
export __orderedExports = { paragraph, heading };
```

if you want to "inject" it into the export default object then you probably need access to a full AST. 

also imagine something like this

```js
const meta = {};
meta.title = 'MyComponent';

export default meta;
```

or even this
```js
import { meta } from './meta.json';

export default meta;
```

it becomes much harder to inject the data into the right place?

Comment (Hypnosphi) [reply lib/core/src/client/preview/start.js]: Oh, I missed the loader part. How are you going to use a loader without webpack BTW?

Comment (Hypnosphi) [reply lib/core/src/client/preview/start.js]: > is handcrafted by users to define specific stories that should be included (and in which order)

Right now the "in which order" part doesn't work with native ES modules

Comment (daKmoR) [reply lib/core/src/client/preview/start.js]: > Oh, I missed the loader part. How are you going to use a loader without webpack BTW?

we are using [es-dev-server](https://open-wc.org/developing/es-dev-server.html) which is based on a koa server (evolvement of express) which can add multiple middlewares where each can process the request before sending it to the client
You can read more about how we use webpack here
https://dev.to/open-wc/storybook-for-web-components-on-steroids-4h29

Also if webpack ever changes the order then storybook will need to add a loader as well

**Review by Hypnosphi — APPROVED** [review]

**Comment by daKmoR** [comment]
> I agree with this sentiment but I just dream of a day where we can drop babel and webpack and all source transformation tools and just load things natively. This seems like a pretty big roadbump in getting there :/

yes indeed... if source order matters then an array (or object) would have been the way too go... but the developer experience when writing stories would have been way lower :/

you could still do it build less by running es-module-lexer in the browser - its performance impact should be fairly minimal as it is extremely efficient both in algorithm and through wasm.

that is however for another time (and there is still much to be done before)

**Comment by tmeasday** [comment]
> if source order matters then an array (or object) would have been the way too go

Yeah, and based on this I predict CSFv2 is going to have just that :)
